### PR TITLE
Tweak parsing recovery of enums, for exprs and match arm patterns

### LINKED ISSUE #100825: Parenthesized match guards should have a specific parser error (and suggestion)
Given the following code: [[playground]](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=50cbb8aafd2bd1cb1bb6f749ee2f1a34)

```rust
pub fn test(val: i32) {
    match val {
        (0 if true) => {}
        _ => {}
    }
}
```

The current output is:

```
error: expected identifier, found keyword `if`
 --> src/lib.rs:3:12
  |
3 |         (0 if true) => {}
  |            ^^ expected identifier, found keyword

error: expected one of `)`, `,`, `...`, `..=`, `..`, or `|`, found keyword `if`
 --> src/lib.rs:3:12
  |
3 |         (0 if true) => {}
  |           -^^ expected one of `)`, `,`, `...`, `..=`, `..`, or `|`
  |           |
  |           help: missing `,`

error: expected one of `)`, `,`, `@`, or `|`, found keyword `true`
 --> src/lib.rs:3:15
  |
3 |         (0 if true) => {}
  |              -^^^^ expected one of `)`, `,`, `@`, or `|`
  |              |
  |              help: missing `,`

error[E0308]: mismatched types
 --> src/lib.rs:3:9
  |
2 |     match val {
  |           --- this expression has type `i32`
3 |         (0 if true) => {}
  |         ^^^^^^^^^^^ expected `i32`, found tuple
  |
  = note: expected type `i32`
            found tuple `(_, _, _)`
```

Ideally the output should look like:

```
error: match pattern guard not allowed here
 --> src/lib.rs:3:9
  |
3 |         (0 if true) => {}
  |            ^^^^^^^ not supported
  |
  = note: match pattern guards must appear at the top level
help: move the guard to the top level
  |
3 -         (_ if true) => {}
3 +         (_) if true => {}
  |
```

As-is, results in `unused_parens` warning. Alternative help that doesn't:

```
help: remove these parentheses
  |
3 -         (0 if true) => {}
3 +         0 if true => {}
  |
```

Bonus points:

- Stash this error while parsing and tweak it later when types are available; e.g. if the scrutinee actually is `(_, _, _)`, perhaps `(_, r#if, true)` (what this currently seemingly error-recovers as) actually *was* meant. With a non-tuple scrutinee, though, clearly the parenthesization was meant.
- Support parsing parenthesized guards in nested patterns and give the nice parser error there. Note that for the guard lifting to be semantics preserving, it must not exit any alt (`|`) pattern combinations.
    <details><summary>examples</summary>

    ```rust
    Some(0 if true) => Some(_) if true;
    (0, 0 if true) => (0, 0) if true;
    (_ if true) | 0 => {} // cannot be written as a single pattern
    0 | _ if true => 0 | _ if true; // is a valid pattern: (0 | _) if true
    Struct { field: 0 if true } => Struct { field: 0 } if true;
    ```
    -----
    </details>

@rustbot label +D-confusing +D-verbose

(Is this https://github.com/rust-lang/rust/labels/D-invalid-suggestion? The `help` is not moving towards correct code.)

### PR DESCRIPTION
**Pull request by estebank**
Tweak recovery of `for (pat in expr) {}` for more accurate spans.

When encountering `match` arm `(pat if expr) => {}`, recover and suggest removing parentheses. Fix #100825.

When encountering malformed enums, try more localized per-variant parse recovery.

Move parser recovery tests to subdirectory.

### CODE REVIEW

File: compiler/rustc_parse/src/parser/expr.rs [hunk 1]
```diff
@@ -2589,30 +2589,66 @@ impl<'a> Parser<'a> {
...
+            let start_span = self.token.span;
+            let left = self.prev_token.span.between(self.look_ahead(1, |t| t.span));
+            match self.parse_pat_allow_top_alt(
+                None,
+                RecoverComma::Yes,
+                RecoverColon::Yes,
+                CommaRecoveryMode::LikelyTuple,
+            ) {
+                Ok(pat) => pat,
+                Err(err) if self.eat_keyword(kw::In) => {
```
Comment (oli-obk) [inline compiler/rustc_parse/src/parser/expr.rs:2606]: Please leave some comments on these code paths explaining what code leads here and what the behaviour is after it

**Review by oli-obk — CHANGES_REQUESTED** [review]

File: tests/ui/parser/issues/issue-104088.rs [hunk 2]
```diff
@@ -8,10 +8,12 @@ fn 1234test() {
 
     let 23name = 123;
     //~^ ERROR expected identifier, found `23name`
-
+}
```
Comment (Nilstrieb) [inline tests/ui/parser/issues/issue-104088.rs:11]: can we move this test to `recovery/recover-number-ident` or something like that? it's not a regression test for a weird case so i think it's pretty misplaced here

File: compiler/rustc_parse/src/parser/expr.rs [hunk 3]
```diff
@@ -2589,30 +2589,66 @@ impl<'a> Parser<'a> {
...
-    /// Parses `for <src_pat> in <src_expr> <src_loop_block>` (`for` token already eaten).
-    fn parse_expr_for(&mut self, opt_label: Option<Label>, lo: Span) -> PResult<'a, P<Expr>> {
-        // Record whether we are about to parse `for (`.
-        // This is used below for recovery in case of `for ( $stuff ) $block`
-        // in which case we will suggest `for $stuff $block`.
-        let begin_paren = match self.token.kind {
-            token::OpenDelim(Delimiter::Parenthesis) => Some(self.token.span),
-            _ => None,
+    fn parse_for_head(&mut self) -> PResult<'a, (P<Pat>, P<Expr>)> {
+        let pat = if self.token.kind == token::OpenDelim(Delimiter::Parenthesis) {
```
Comment (Nilstrieb) [inline compiler/rustc_parse/src/parser/expr.rs:2593]: instead of branching on this here, can we keep the previous code that saved it into a local and then calls `parse_path_allow_top_alt`? you can then match on the result like in the if case and just return the error if there's no `(`.
that makes it more obvious that the `(` only matters in the error case, which is important

**Review by Noratrieb — CHANGES_REQUESTED** [review]
This PR is quite big and changes a lot of things, making it hard to follow and doesn't allow merging the parts that are good already. It also makes it harder to see which test changes are caused by which code changes. Can you split it up into several PRs for the different parts, like enums, for loops and match arms?

File: compiler/rustc_parse/src/parser/expr.rs (hunk 3 above)
Comment (estebank) [reply compiler/rustc_parse/src/parser/expr.rs]: I think I got what you meant, check the last commit to verify I did.

File: compiler/rustc_parse/src/parser/expr.rs (hunk 1 above)
Comment (estebank) [reply compiler/rustc_parse/src/parser/expr.rs]: Added a comment, let me know if it is clearer with it.

**Comment by estebank** [comment]
I split the commits as much as possible. I didn't bother making the test comments be accurate for each commit, but did include the `.stderr` changes corresponding to each of them.

**Review by Noratrieb — COMMENTED** [review]
Thanks for splitting, that makes it a lot easier to review. I've already reviewed the first 4 commits and they look good. I'll review the rest later.
If you want those merged already (making your job rebasing this easier) you can split them into a separate PR and r=me it. Or you can wait here if you prefer.

File: compiler/rustc_parse/src/parser/expr.rs [hunk 4]
```diff
@@ -3023,6 +2986,87 @@ impl<'a> Parser<'a> {
...
+                self.sess.gated_spans.ungate_last(sym::let_chains, cond.span);
+            }
+            let span = if_span.to(cond.span);
+            self.sess.gated_spans.gate(sym::if_let_guard, span);
+        }
+        Ok(Some(cond))
+    }
+
+    fn parse_match_arm_pat_and_guard(&mut self) -> PResult<'a, (P<Pat>, Option<P<Expr>>)> {
+        if self.token.kind == token::OpenDelim(Delimiter::Parenthesis) {
```
Comment (Nilstrieb) [inline compiler/rustc_parse/src/parser/expr.rs:3066]: To make it obvious that the only semantic difference is in the error path, I'd prefer to move the call to `parse_pat_allow_top_alt` out of the `if` and move the if down to the error case.
This doesn't have to block the PR and can be done as a follow-up (it wouldn't be too bad if it never happens either, but it would be nice to have it).`

File: compiler/rustc_parse/src/parser/expr.rs (hunk 3 above)
Comment (Nilstrieb) [reply compiler/rustc_parse/src/parser/expr.rs]: you did, thanks!

File: compiler/rustc_parse/src/parser/expr.rs [hunk 5]
```diff
@@ -1,3 +1,4 @@
+// ignore-tidy-filelength
```
Comment (Nilstrieb) [inline compiler/rustc_parse/src/parser/expr.rs:1]: :(

(not really something this PR has to fix though, but it may be nice to split this file up further in the future, maybe by moving out more diagnostic paths)

**Review by Noratrieb — APPROVED** [review]
finally got around to review it fully, thanks for the patience and sorry for the delay, partially missing time and partially also just forgetting about it :skull:

one suggestion for an improvement but that can be a follow up

Splitting up the PR into the many commits was definitely very helpful, thanks. For the future, I think you could have gone further and split every semantic change "pattern changes", "for loop changes", "enum changes", would have also made it easier to see the final UI test diff.

**Comment by Nilstrieb** [comment]
Didn't follow all the recovery paths super closely, but I think I got the high level idea and it looks good. Specifically looked out for happy-path regressions, doesn't look like it had any.
So with that
@bors r+

**Comment by matthiaskrgr** [comment]
@bors r-
may need rebase https://github.com/rust-lang/rust/pull/118432#issuecomment-1831165864 :/
